---
title: DL1001
description: The description for analyzer rule DL1001
---
## EventsMustBeImmutable

Events that implement the interface `Dolittle.Events.IEvent` are considered
to be immutable events. This means that it is expected that any implementations
of these does not hold a property that has a public setter.

This means that any initial state needs to be given to a constructor that sets
the property.

Since events are serialized and deserialized, the serializers must be able to
deserialize instances. In order for it to be able to do that, it is expecting
the constructor arguments to match the properties. It takes into account naming
conventions, and expects the arguments in the constructor to be [lower camel case](https://en.wikipedia.org/wiki/Camel_case)
while typically the properties are [upper camel case or pascal case](https://en.wikipedia.org/wiki/Camel_case).

Example of a wrong event:

```csharp
public class MyEvent : IEvent
{
    public string MyProperty { get; set; }
}
```

Example of a correct implementation:

```csharp
public class MyEvent : IEvent
{
    public MyEvent(string myProperty)
    {
        MyProperty = myProperty;
    }

    public string MyProperty { get; }
}
```
